snap/info_snap_yaml.go

```diff
@@ -503,9 +503,23 @@ func bindUnboundSlots(snap *Info) {
 
 // bindImplicitHooks binds all global plugs and slots to implicit hooks
 func bindImplicitHooks(snap *Info) {
+	bound := make(map[string]bool, len(snap.Plugs)+len(snap.Slots))
 	for _, plug := range snap.toplevelPlugs {
-		for hookName, hook := range snap.Hooks {
-			if hook.Explicit {
+		if len(plug.Apps) > 0 || len(plug.Hooks) > 0 {
+			bound[plug.Name] = true
+		}
+	}
+	for _, slot := range snap.toplevelSlots {
+		if len(slot.Apps) > 0 || len(slot.Hooks) > 0 {
```

Comment:
There is still something not quite right I fear going on, it's a bit too late to check this here, which means again missing tests.  What we care is really the state of plugs and slots before we run bindUnboundPlugs/Slots.  Maybe the easiest is to replace toplevelPlugs/Slots with unboundPlugs/Slots and compute them after  setHooksFromSnapYaml.

Reply:
I agree, I see the mistake now.

The mistake is this (symmetric for plus):
- a plug is defined at the top level
- it is not referenced anywhere so it binds to all apps and hooks
- it gets bound to an app or a hook and thus the referenced condition no longer holds true
- it does not get bound to an implicitly defined hook, **incorrectly so**

Reply:
This is now fixed.